aspnet/fundamentals/dependency-injection.rst

```diff
@@ -1,10 +1,291 @@
+
+When a system is designed to use DI, with many classes requesting their dependencies via their constructor (or properties), it's helpful to have a class dedicated to creating these classes with their associated dependencies. These classes are referred to as *containers*, or more specifically, `Inversion of Control (IoC) <http://deviq.com/inversion-of-control/>`_ containers or Dependency Injection (DI) containers. A container is essentially a factory that is responsible for providing instances of types that are requested from it. If a given type has declared that it has dependencies, and the container has been configured to provide the dependency types, it will create the dependencies as part of creating the requested instance. In this way, complex dependency graphs can be provided to classes without the need for any hard-coded object construction. In addition to creating objects with their dependencies, containers typically manage object lifetimes within the application.
+
+ASP.NET 5 includes a simple built-in container (represented by the IServiceProvider interface) that supports constructor injection, and ASP.NET makes certain services available through DI. ASP.NET's container refers to the types it manages as *services*. Throughout the rest of this article, *services* will refer to types that are managed by ASP.NET 5's IoC container. You configure the built-in container's services in the ``ConfigureServices`` method in your application's ``Startup`` class.
+
+.. note:: Martin Fowler has written an extensive article on `Inversion of Control Containers and the Dependency Injection Pattern <http://www.martinfowler.com/articles/injection.html>`_. Microsoft Patterns and Practices also has a great description of `Dependency Injection <https://msdn.microsoft.com/en-us/library/dn178469(v=pandp.30).aspx>`_.
+
+Using Framework-Provided Services
+---------------------------------
+
+ASP.NET 5 provides certain framework services and objects used to communicate with the framework during your application's startup. You can request any of these services or objects by simply including the appropriate interface as a parameter on your ``Startup`` class's constructor or one of its ``Configure`` or ``ConfigureServices`` methods. These services include:
```

Comment:
> certain framework services and objects ... or one of its ``Configure`` or ``ConfigureServices`` methods

Double-checked and this is a great way to see `InvalidOperationException`s. The `ConfigureServices()` method has a few supported signatures e.g.
``` c#
    void ConfigureServices();
    IServiceProvider ConfigureServices();
    void ConfigureProductionServices(IServiceProvider services);
```
But any parameter or non-void return that isn't of type `IServiceProvider` leads to an exception.

On the other hand, there is nothing special about "framework" services as `Configure` parameters. Users can add parameters of _any_ registered service type to the `Configure` method. If this needs to be described here and not in the `Startup` topic, cover the constructor -> `ConfigureServices()` -> `Configure()` ordering and mention the constructor parameters are unusual w.r.t. available services -- precisely because `ConfigureServices()` hasn't been called when the constructor is invoked. (FYI and not for the article, I successfully tested a `Configure()` method like the following with only `AddMvc()` in the `ConfigureServices()` method.)
``` c#
       public void Configure(
            IApplicationBuilder app,
            IHostingEnvironment env,
            ILoggerFactory loggerFactory,
            IHtmlHelper helper,
            IEnumerable<IConfigureOptions<MvcOptions>> configurators)
```

> These services include:

"services and objects"